### PR TITLE
feat(new source): new `azure_event_hubs` source.

### LINKED ISSUE #4261: New `azure_event_hubs` source
Ability so source data from [Azure Event Hubs](https://azure.microsoft.com/en-us/services/event-hubs/).

### PR DESCRIPTION
**Pull request by StephenWakely**
Closes #4261 

This adds a new source that accepts data from [Azure event hubs](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-about).

It is a simple wrapper over the Kafka source since Event Hubs provides a [Kafka compatible API](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview).

It should be noted that the Kafka connection doesn't work with the Basic SKU. Users will need to have a Standard, Premium and Dedicated SKU in order to use this source.

Currently this exposes a minimum set of options that are passed to the Kafka source. I am open to adding more - for example access to the `librdkafka` options if it is deemed that this will be useful.

### CODE REVIEW

File: src/sources/azure_event_hubs/config.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        default_session_timeout_ms, default_socket_timeout_ms, default_topic_key,
+        KafkaSourceConfig,
+    },
+    tls::TlsEnableableConfig,
+};
+
+/// Configuration for the `azure_event_hubs` source.
+/// This component is a simple wrapper over the `kafka` source.
+/// See the documentation [here](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview)
+/// for details on how `azure_event_hubs` can use `kafka`.
```
Comment (spencergilbert) [inline src/sources/azure_event_hubs/config.rs:22]:
```suggestion
/// See the documentation [here][event_hubs_docs] for details on how `azure_event_hubs` can use `kafka`.
///
///[event_hubs_docs]: https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview
```

I'd use this format instead - [ref](https://github.com/vectordotdev/vector/blob/master/src/sinks/kafka/config.rs#L67-L71)

File: src/sources/azure_event_hubs/config.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        default_auto_offset_reset, default_commit_interval_ms, default_fetch_wait_max_ms,
+        default_headers_key, default_key_field, default_offset_key, default_partition_key,
+        default_session_timeout_ms, default_socket_timeout_ms, default_topic_key,
+        KafkaSourceConfig,
+    },
+    tls::TlsEnableableConfig,
+};
+
+/// Configuration for the `azure_event_hubs` source.
+/// This component is a simple wrapper over the `kafka` source.
```
Comment (spencergilbert) [inline src/sources/azure_event_hubs/config.rs:20]: This probably doesn't need to be included as we wouldn't necessarily want it rendered in the docs?

File: src/sources/azure_event_hubs/config.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+/// See the documentation [here](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview)
+/// for details on how `azure_event_hubs` can use `kafka`.
+#[configurable_component(source)]
+#[derive(Clone, Debug, Derivative)]
+#[derivative(Default)]
+#[serde(deny_unknown_fields)]
+pub struct AzureEventHubsConfig {
+    /// The connection string.
+    /// See [here](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-get-connection-string) for details.
```
Comment (spencergilbert) [inline src/sources/azure_event_hubs/config.rs:29]:
```suggestion
    /// See [here][connection_string] for details.
    ///
    /// [connection_string]: https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-get-connection-string
```

File: website/cue/reference/components/sources/azure_event_hubs.cue [hunk 4]
```diff
@@ -0,0 +1,102 @@
+package metadata
+
+components: sources: azure_event_hubs: {
+	title: "Azure Event Hubs"
+
+	classes: {
+		commonly_used: false
+		delivery:      "at_least_once"
+		deployment_roles: ["daemon", "sidecar"]
```
Comment (spencergilbert) [inline website/cue/reference/components/sources/azure_event_hubs.cue:9]: Not aggregator?

File: website/cue/reference/components/sources/azure_event_hubs.cue [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		development:   "beta"
+		egress_method: "batch"
+		stateful:      false
+	}
+
+	features: {
+		acknowledgements: false
```
Comment (spencergilbert) [inline website/cue/reference/components/sources/azure_event_hubs.cue:16]:
```suggestion
		acknowledgements: true
```

File: website/cue/reference/components/sources/azure_event_hubs.cue [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		}
+	}
+
+	how_it_works: {
+		kafka: {
+			title: "kafka"
+			body:  """
+                This component is a simple wrapper over the `kafka` source.
+                See the documentation [here](\(urls.azure_event_hubs_kafka_ecosystem))
+                for details on how `azure_event_hubs` can use `kafka`.
```
Comment (spencergilbert) [inline website/cue/reference/components/sources/azure_event_hubs.cue:85]: I'd rather have this say something like "this component leverages event hubs' compatibility with kafka and requires <whatever account level>..."

File: Cargo.toml [hunk 7]
```diff
@@ -487,6 +488,7 @@ sources-aws_ecs_metrics = []
 sources-aws_kinesis_firehose = ["dep:base64", "dep:infer", "sources-utils-tls"]
 sources-aws_s3 = ["aws-core", "dep:aws-sdk-sqs", "dep:aws-sdk-s3", "dep:semver", "dep:async-compression", "sources-aws_sqs", "tokio-util/io"]
 sources-aws_sqs = ["aws-core", "dep:aws-sdk-sqs"]
+sources-azure_event_hubs = []
```
Comment (spencergilbert) [inline Cargo.toml:491]: Don't need to grab the kafka features here?

**Review by spencergilbert — CHANGES_REQUESTED** [review]
Mostly issues with the docs

File: website/cue/reference/components/sources/azure_event_hubs.cue [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+				"""
+		}
+	}
+
+	telemetry: metrics: {
```
Comment (jszwedko) [inline website/cue/reference/components/sources/azure_event_hubs.cue:90]: Do we need to add the event processing metrics, like `received_event_bytes_total` here too?

**Review by jszwedko — COMMENTED** [review]
Sweet. This generally looks good to me. Agreed with Spencer's comments on the docs. For rdkafka options, I suggest we wait until someone asks for them.

One thought I did have is that we'll have to refactor this configuration quite a bit if we do end up wanting to support other ways of pulling data from Azure Event Hubs (like their AMQP API).

**Comment by spencergilbert** [comment]
> One thought I did have is that we'll have to refactor this configuration quite a bit if we do end up wanting to support other ways of pulling data from Azure Event Hubs (like their AMQP API).

🤔 what's the likelihood of this, we should get ahead of it if it seems at all likely (ex `opentelemetry` logs, metrics, traces.

**Comment by jszwedko** [comment]
> > One thought I did have is that we'll have to refactor this configuration quite a bit if we do end up wanting to support other ways of pulling data from Azure Event Hubs (like their AMQP API).
> 
> 🤔 what's the likelihood of this, we should get ahead of it if it seems at all likely (ex `opentelemetry` logs, metrics, traces.

I'm not really sure when someone would need / want to use a specific interface. My gut says it is a low risk, I think we may just need to wait for users to ask for it to understand the use-cases better.

File: src/sources/azure_event_hubs/mod.rs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+mod config;
+
+pub use config::AzureEventHubsConfig;
+
+use crate::config::SourceDescription;
+
+inventory::submit! {
+    SourceDescription::new::<AzureEventHubsConfig>("azure_event_hubs")
+}
```
Comment (neuronull) [inline src/sources/azure_event_hubs/mod.rs:9]: file is missing a newline

**Review by neuronull — APPROVED** [review]

**Comment by jeromekleinen-kbc** [comment]
I hope you guys don't mind if I chime in here. I have been using the kafka source to read AEH for quite some time now. What is the advantage of this dedicated source compared to just providing a specific guide? Or in other words: should I switch my existing configs when this lands?

I quickly skimmed the code and I think no specific librdkafka options are set. I have been using the ones specified by Microsoft here: https://github.com/Azure/azure-event-hubs-for-kafka/blob/master/CONFIGURATION.md#librdkafka-configuration-properties

I honestly don't know whether they make any real difference though.

**Comment by spencergilbert** [comment]
> I hope you guys don't mind if I chime in here. I have been using the kafka source to read AEH for quite some time now. What is the advantage of this dedicated source compared to just providing a specific guide? Or in other words: should I switch my existing configs when this lands?
> 
> I quickly skimmed the code and I think no specific librdkafka options are set. I have been using the ones specified by Microsoft here: https://github.com/Azure/azure-event-hubs-for-kafka/blob/master/CONFIGURATION.md#librdkafka-configuration-properties
> 
> I honestly don't know whether they make any real difference though.

Of course we don't mind 😄 it's definitely an ease of config situation. If the `kafka` source is working fine for you there isn't any "secret sauce" in the new source. Think of it more like "syntactic sugar" for the config.

@StephenWakely, I'm not sure we knew they had recommended values for `librdkafka` - I think that gives a good reason to expose that in the config (if not default to them).

**Comment by jszwedko** [comment]
@jeromekleinen-kbc to add on to what Spencer said, you can find some more detail in: https://github.com/vectordotdev/vector/blob/master/docs/USER_EXPERIENCE_DESIGN.md#source--sink-boundaries. Essentially having a dedicated sink helps with discoverability and streamlines user configuration.

**Comment by StephenWakely** [comment]
The Kafka protocol doesn't work with the Azure basic tier. A proper solution needs to use the AMQP protocol, which we will look at doing in due course. Closing this.

**Pull request closed without merging**